File: ProgettoKalk/KalkJava/Cerchio.java
import java.lang.Math;
public class Cerchio extends Figura{
    Punto centro;
    double raggio;
    static int idCerchio=1;

    public Cerchio(double x, double y, double r){
        super("Cerchio " + idCerchio);
        centro= new Punto(x,y);
        raggio=r;
        setNome("Cerchio" + idCerchio);
        idCerchio++;
    }
    public Punto getCentro(){
        return centro;
    }
    public double getRaggio(){
        return raggio;
    }
    public String toString() {
        return getNome() + ": Centro " + centro.toString() + "Raggio " + raggio;
    }
    public double distanza(Figura f) {
        double d, d1;
        if(f instanceof Segmento){
            Segmento s= (Segmento)f;
            d=Punto.distanza(centro,s.getA());
            d1=Punto.distanza(centro,s.getB());
            if(d<d1) return d;
            else return d1;
        }
        if(f instanceof Poligono) {
            Poligono p = (Poligono) f;
            d = Punto.distanza(centro, p.baricentro());
            return d;
        }
        if(f instanceof Cerchio){
            Cerchio c= (Cerchio)f;
            d=Punto.distanza(centro,c.centro);
            return d;
        }
        return -1;
    }
    public double area(){
        return Math.PI*Math.pow(raggio,2);
    }
    public double perimetro(){
        return (Math.PI*raggio)*2;
    }
    public double diametro(){
        return (2*raggio);
    }
}

File: ProgettoKalk/KalkMV/Model/cerchio.cpp
#include "cerchio.h"
using namespace std;

unsigned int Cerchio::idCerchio=1;

Cerchio::Cerchio(double x, double y, double r):
    Figura("Cerchio "+std::to_string(idCerchio)),centro(x,y), raggio(r){
    idCerchio++;
}

Punto Cerchio::getCentro() const {
    return centro;
}

double Cerchio::perimetro() const {
    return (M_PI*raggio)*2;
}

double Cerchio::area() const{
    return M_PI*pow(raggio,2);
}

double Cerchio::distanza(Figura* pf) const{
    double d, d1;
    Segmento* ps = dynamic_cast<Segmento*>(pf);
    if(ps){
        d=Punto::distanza(centro,ps->getA());
        d1=Punto::distanza(centro,ps->getB());
        if(d<d1) return d;
        else return d1;
    }
    Poligono* pp=dynamic_cast<Poligono*>(pf);
    if(pp){
        d=Punto::distanza(centro,pp->baricentro());
        return d;
    }
    Cerchio* pc=dynamic_cast<Cerchio*>(pf);
    if(pc){
        d=Punto::distanza(centro,pc->centro);
        return d;
    }
}

string Cerchio::print() const{
    return "Centro: "+ centro.print()+ "\nraggio: " + std::to_string(int(raggio));
}

Cerchio::~Cerchio()
{

}
